File: Java/Builder/LaptopManufacturer/src/com/ericgtkb/LaptopBuilder.java
package com.ericgtkb;

public interface LaptopBuilder {
    void setLaptopName();
    void installCpu();
    void installMemory();
    void installOs();
    Laptop getLaptop();
}

class MacbookBuilder implements LaptopBuilder {
    private Laptop laptop;

    public MacbookBuilder() {
        laptop = new Laptop();
    }

    @Override
    public void setLaptopName() {
        laptop.setName("Macbook");
    }

    @Override
    public void installCpu() {
        laptop.setCpu("Intel i7");
    }

    @Override
    public void installMemory() {
        laptop.setMemory("16 GB");
    }

    @Override
    public void installOs() {
        laptop.setOs("macOS");
    }

    @Override
    public Laptop getLaptop() {
        return laptop;
    }
}

class LinuxLaptopBuilder implements LaptopBuilder {
    private Laptop laptop;

    public LinuxLaptopBuilder() {
        laptop = new Laptop();
    }

    @Override
    public void setLaptopName() {
        laptop.setName("Linux laptop");
    }

    @Override
    public void installCpu() {
        laptop.setCpu("Intel i9");
    }

    @Override
    public void installMemory() {
        laptop.setMemory("32 GB");
    }

    @Override
    public void installOs() {
        laptop.setOs("linux (Ubuntu)");
    }

    @Override
    public Laptop getLaptop() {
        return laptop;
    }
}


class WindowsLaptopBuilder implements LaptopBuilder {
    private Laptop laptop;

    public WindowsLaptopBuilder() {
        laptop = new Laptop();
    }

    @Override
    public void setLaptopName() {
        laptop.setName("Windows laptop");
    }

    @Override
    public void installCpu() {
        laptop.setCpu("Intel i5");
    }

    @Override
    public void installMemory() {
        laptop.setMemory("16 GB");
    }

    @Override
    public void installOs() {
        laptop.setOs("windows 10");
    }

    @Override
    public Laptop getLaptop() {
        return laptop;
    }
}

File: C++/Builder/LaptopManufacturer/laptop_builder.cc
#include "laptop_builder.h"

MacbookBuilder::MacbookBuilder() : _laptop{std::make_unique<Laptop>()} {
}

void MacbookBuilder::set_laptop_name() {
    _laptop->set_name("Macbook");
}

void MacbookBuilder::install_cpu() {
    _laptop->set_cpu("Intel i7");
}

void MacbookBuilder::install_memory() {
    _laptop->set_memory("16 GB");
}

void MacbookBuilder::install_os() {
    _laptop->set_os("macOS");
}

std::unique_ptr<Laptop> MacbookBuilder::get_laptop() {
    return std::move(_laptop);
}


LinuxLaptopBuilder::LinuxLaptopBuilder() : _laptop{std::make_unique<Laptop>()} {
}

void LinuxLaptopBuilder::set_laptop_name() {
    _laptop->set_name("Linux laptop");
}

void LinuxLaptopBuilder::install_cpu() {
    _laptop->set_cpu("Intel i9");
}

void LinuxLaptopBuilder::install_memory() {
    _laptop->set_memory("32 GB");
}

void LinuxLaptopBuilder::install_os() {
    _laptop->set_os("linux (Ubuntu)");
}

std::unique_ptr<Laptop> LinuxLaptopBuilder::get_laptop() {
    return std::move(_laptop);
}


WindowsLaptopBuilder::WindowsLaptopBuilder() : _laptop{std::make_unique<Laptop>()} {
}

void WindowsLaptopBuilder::set_laptop_name() {
    _laptop->set_name("Window laptop");
}

void WindowsLaptopBuilder::install_cpu() {
    _laptop->set_cpu("Intel i5");
}

void WindowsLaptopBuilder::install_memory() {
    _laptop->set_memory("16 GB");
}

void WindowsLaptopBuilder::install_os() {
    _laptop->set_os("windows 10");
}

std::unique_ptr<Laptop> WindowsLaptopBuilder::get_laptop() {
    return std::move(_laptop);
}
